Proposal for moving data between two JBOD disks using Cruise Control (#106)
strimzi/proposals: 063-moving-data-between-two-jbod-disks-using-cruise-control.md

# Moving data between two JBOD disks using Cruise Control

> This proposal is about integrating the `remove_disks` endpoint from Cruise Control into Strimzi cluster operator. 

Reviewer: I can't find any documentation about this endpoint in Cruise Control, for example it's missing in the wiki.
Is there an official place where we can find how it works and put a link here?

Reviewer 2: Do we want to link to the [yaml](https://github.com/ilievladiulian/cruise-control/blob/ad275079f271b28704a1d5ff00fff82891aebc89/cruise-control/src/main/resources/yaml/endpoints/removeDisks.yaml)

Author: I will create a documentation bug and create a PR for the same on upsteam repo regarding this. The only piece of information for using this is only on the PR

> This endpoint will allow us to move the data between two JBOD disks. 

Reviewer 3: Does it only allow moving data between two disks?

## Current situation

> Currently, we get a multiple requests to add the ability for moving all Kafka logs between two disks on the JBOD storage array. This feature can be useful in following scenarios:

Reviewer: multiple requests ... from community users I guess? Let's make it clear.

> Currently, we get a multiple requests to add the ability for moving all Kafka logs between two disks on the JBOD storage array. This feature can be useful in following scenarios:

Reviewer 3: Would this be better in the  `Motivation`section?

- The current disk is too big and the user wants to use smaller one
- When we want to use different Storage Class with different parameters or different storage types.

> For now, we can do this using the Kafka CLI tools but it is not very user-friendly.

Reviewer: which Kafka CLI tools?

Reviewer 2: I guess we want to mention `kafka-reassign-partitions.sh` tool

> For now, we can do this using the Kafka CLI tools but it is not very user-friendly.

Reviewer 3: This is a reference to the `kafka-reassign-partitions` script right? If so, you could emphasize how it is devising the partition mappings to migrate data between disks in this scenario is a manual, time-consuming process. Solutions exist for moving data of brokers but not off specific disks of brokers which have JBOD config

```
where:
```sh
remainingUsageAfterRemoval = current usage for remaining disks + additional usage from removed disks
remainingCapacity = sum of capacities of the remaining disks
errorMargin = configurable property (default 0.1); it makes sure that a disk percentage is always free when moving replicas
```

In order to use the above endpoint in the Strimzi cluster operator, it would be added to the [`CruiseControlApi`](https://github.com/strimzi/strimzi-kafka-operator/blob/main/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/cruisecontrol/CruiseControlApi.java) interface and developing the corresponding implementation.

### Implementation

For implementing this feature, We will be adding a new mode to the `KafkaRebalanceMode` class.
* `remove_disks`: It moves replicas from a specified disk to other disks of the same broker
You can use this mode by changing the `spec.mode` to `remove_disks` in the `KafkaRebalance` resource.

A `KafkaRebalance` custom resource would look like this.

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaRebalance
metadata:
  name: my-rebalance
  labels:
    strimzi.io/cluster: my-cluster
spec:
  mode: remove-disks

>   brokerandlogdirs: [brokerid1-logdir1,brokerid2-logdir2...] # for eg. 0-/var/lib/kafka/data-0/kafka-log0  

Reviewer 4: `brokerandlogdirs` seems strange as a name. If nothing else, you should use proper case: `brokerAndLogDirs`. For the users, it is quite complicated to understand what the paths are etc. as they are an internal implementation. So I also wonder if you should create more elegant APIs. E.g. something like:
```
spec:
  mode: remove-disks
  brokerandlogdirs:
    - brokerId: 0
      volumeId: 0
    - brokerId: 1
      volumeId: 0
    - brokerId: 2
      volumeId: 0
```

Or something similar that would be based on what the user configures int he Kafka / KafkaNodePool resources and nto on the internal paths.

Reviewer: I agree with Jakub. In our current Strimzi implementation, the users have no knowledge of the internal paths on the volume where we configure the brokers to store logs. The only thing the user knows is the volume on a specific broker as pointed out  in the above example.
The user doesn't know about `/var/lib/kafka/data-0/kafka-log0` at all but we are using a very well defined pattern to make it depending on the volume for a broker.

# ...
```

### Flow

> - The user should be using the `Kafka` resource with JBOD configured. 

Reviewer 4: What if the user uses KafkaNodePool resources? What if the user does not use JBOD type storage?

> - When the Kafka cluster is ready, the user creates a `KafkaRebalance` custom resource with the `spec.mode` field as `remove-disks` and the list of the broker and the corresponding logdirs to move in the `spec.brokerandlogdirs` field.

Reviewer 5: You have `remove_disks` on one place and `remove-disks` one another - it should be the same across whole proposal

> - The `KafkaRebalanceAssemblyOperator` starts the interaction with Cruise Control via the `/remove_disks` endpoint for getting an optimization proposal (by using the dryrun feature). 

Reviewer 2: ```suggestion
- The `KafkaRebalanceAssemblyOperator` interacts with Cruise Control via the `/remove_disks` endpoint to generate an optimization proposal (by using the dryrun feature). 
```

> - The user accepts the proposal by applying the `strimzi.io/rebalance=approve` annotation on it. 

Reviewer: I guess the user can still use the auto approval feature we have. It's better to highlight it here.

> - The `KafkaRebalanceAssemblyOperator` starts the interaction with Cruise Control via the `/remove_disks` endpoint for running the actual rebalancing.

Reviewer 2: ```suggestion
- The `KafkaRebalanceAssemblyOperator` interacts with Cruise Control via the `/remove_disks` endpoint to perform the actual rebalancing.
```

> Note: The movement of data between the JBOD disks doesn't affect the broker load, therefore there will be no changes in the before/after broker load.

Reviewer 4: Why does it not impact it? The moving of data between disks has to impact performance and load of the broker.

Reviewer 3: I mislead Shubham on this one, I thought it wouldn't affect the broker load since the data going in and out of the broker would be the same regardless of how the data was distributed amongst the disks. But I understand that if one disk was faster than another, moving the data between them within a broker certainly would impact the performance 

@ShubhamRwt did we ever find out why the loadBefore/loadAfter information was not provided by this `remove_disk` endpoint? Is it just an oversight by upstream Cruise Control?
